File: dart/collections/maps.dart
import 'dart:collection';

void main() {
  // Empty Map
  var intMap0 = {};
  Map intMap01 = <int, int>{};
  Map<int, int> intMap02 = <int, int>{};

  // Filled Lists
  var intMap1 = {1: 2, 2: 3, 3: 4};
  Map intMap12 = <int, int>{1: 2, 2: 3, 3: 4};
  Map<int, int> intMap13 = <int, int>{1: 2, 2: 3, 3: 4};

  var c = UnmodifiableMapView(intMap1); // Modifiable via seed (intMap1)
  var d = Map.unmodifiable(intMap1); // Rigid unmodifiable

  print("UnmodifiableMapView $c");
  print("Map.unmodifiable $d");

  intMap1.putIfAbsent(100, () => 200);

  print("UnmodifiableListView $c");
  print("Map.unmodifiable $d");

  // Most common use caes
  intMap13.length; // Returns the number of entries of the map.
  intMap13.entries; // Returns an Iterable of entries of the map.
  intMap13.isEmpty; // Returns if the map is empty.
  intMap13.isNotEmpty; // Returns if the map is not empty.
  intMap13.keys; // Returns the list of keys of the map.
  intMap13.values; // Returns the list of values of the map.
  intMap13.addAll({}); // Adds all elements of another map inside the map.
  intMap13.clear(); // Clear the map.
  intMap13.containsKey(1); // Determines if the map contains a key;
  intMap13.containsValue(1); // Determines if the map contains a value;
  intMap13.map((k, v) => MapEntry(v, k)); // Interates over the map entries and returns a new map.
  intMap13.remove(1); // Remove an element which key is equal to 1.
  intMap13.removeWhere((k, v) => k == v); // Remove an element which key is equal to value.
  intMap13.putIfAbsent(3, () => 43); // Insert a value if the does not exists yet.
  intMap13[1]; // Returns the value of key 1.
  intMap13[1] = 678; // Updates the map value for the entry with key equal to 1.
}

File: kotlin/collections/maps.kt
import kotlin.collections.mutableMapOf

fun main() {
    var immutableIntMap0 = mapOf<Int, Int>() // Immutable empty map
    val intMap0 = mutableMapOf<Int, Int>() // Mutable empty map

    val immutableIntMap1 = mapOf(1 to 2, 2 to 3, 3 to 4) // Immutable map
    val intMap1 = hashMapOf(1 to 2, 2 to 3, 3 to 4) // Mutable map

    // it also could be declared as mutableMapOf
    // val intMap1 = hashMapOf(1 to 2, 2 to 3, 3 to 4) // Mutable map

    // Most common use caes
    intMap1.size // Number of entries in the map
    intMap1.entries // Returns a Set of entries of the map.
    intMap1.isEmpty() // Returns if the map is empty.
    intMap1.isNotEmpty() // Returns if the map is not empty.
    intMap1.keys // Returns the keys of the map
    intMap1.values // Returns the values of the map
    intMap1.containsKey(1) // Checks if key 1 is present
    intMap1.containsValue(2) // Checks if value 2 is present
    intMap1.remove(1) // Removes the entry with key 1
    intMap1.remove(3, 4) // Remove the entrey where the key and value matches
    intMap1.putIfAbsent(3, 43) // Insert a value if the does not exists yet.
    intMap1.get(2) // Returns the value of key 2.
    intMap1[2] // Returns the value of key 2.
    intMap1[1] = 678 // Updates value for key 1
    intMap1.clear() // Clears the map
}
